fe/estimator.py

```diff
@@ -1,170 +1,28 @@
-    du_dp_obs = []
-
-    for bp in final_potentials:
-        impl = bp.bound_impl(np.float32)
-        if isinstance(bp, potentials.Nonbonded):
-            nonbonded_impls.append(impl)
-        else:
-            bonded_impls.append(impl)
-        all_impls.append(impl)
-        du_dp_obs.append(custom_ops.AvgPartialUPartialParam(impl, 5))
-
-    if integrator.seed == 0:
-        integrator = copy.deepcopy(integrator)
-        integrator.seed = np.random.randint(np.iinfo(np.int32).max)
-
-    intg_impl = integrator.impl()
-    # context components: positions, velocities, box, integrator, energy fxns
-    ctxt = custom_ops.Context(
-        x0,
-        v0,
-        box,
-        intg_impl,
-        all_impls
-    )
-
-    # equilibration
-    equil_schedule = np.ones(equil_steps)*lamb
-    ctxt.multiple_steps(equil_schedule)
-
-    for obs in du_dp_obs:
-        ctxt.add_observable(obs)
-
-    prod_schedule = np.ones(prod_steps)*lamb
-
-    full_du_dls, xs = ctxt.multiple_steps(prod_schedule, du_dl_interval, x_interval)
-
-    # keep the structure of grads the same as that of final_potentials so we can properly
-    # form their vjps.
-    grads = []
-    for obs in du_dp_obs:
-        grads.append(obs.avg_du_dp())
-
-    result = SimulationResult(xs=xs, du_dls=full_du_dls, du_dps=grads)
-    return result
-
-
 FreeEnergyModel = namedtuple(
     "FreeEnergyModel",
-    ["unbound_potentials", "client", "box", "x0", "v0", "integrator", "lambda_schedule", "equil_steps", "prod_steps"]
+    ["dGs", "dG_grads"]
 )
 
 gradient = List[Any] # TODO: make this more descriptive of dG_grad structure
 
 def _deltaG(model, sys_params) -> Tuple[Tuple[float, List], np.array]:
 
-    assert len(sys_params) == len(model.unbound_potentials)
-
-    bound_potentials = []
-    for params, unbound_pot in zip(sys_params, model.unbound_potentials):
-        bp = unbound_pot.bind(np.asarray(params))
-        bound_potentials.append(bp)
-
-    if model.client is None:
-        results = []
-        for lamb in model.lambda_schedule:
-            results.append(simulate(lamb, model.box, model.x0, model.v0, bound_potentials, model.integrator, model.equil_steps, model.prod_steps))
-    else:
-        futures = []
-        for lamb in model.lambda_schedule:
-            args = (lamb, model.box, model.x0, model.v0, bound_potentials, model.integrator, model.equil_steps, model.prod_steps)
-            futures.append(model.client.submit(simulate, *args))
-
-        results = [x.result() for x in futures]
-
-    mean_du_dls = []
-    all_grads = []
-
-    for result in results:
-        # (ytz): figure out what to do with stddev(du_dl) later
-        mean_du_dls.append(np.mean(result.du_dls))
-        all_grads.append(result.du_dps)
-
-    dG = np.trapz(mean_du_dls, model.lambda_schedule)
     dG_grad = []
-    for rhs, lhs in zip(all_grads[-1], all_grads[0]):
+    for lhs, rhs in zip(model.dG_grads[0][0], model.dG_grads[-1][-1]):
```

Comment:
Document expected nesting of `model.dG_grads`

Reply:
willfix